### ISSUE TITLE
Misleading exception message from `pathlib.PurePath()` when mixing str and bytes arguments

### ISSUE THREAD

**Issue by barneygale** [python/cpython#103525]
Since 6716254e71eeb4666fd6d1a13857832caad7b19f, attempting to create a `pathlib.PurePath` or `Path` object with mixed `str` and `bytes` arguments raises a TypeError (yay!) with a misleading message (booo!):

```python
>>> import pathlib
>>> pathlib.Path('foo', b'bar')
TypeError: Can't mix strings and bytes in path components
```

This message implies that bytes are supported, as long as we don't mix them with strings. And yet when we try that:

```python
>>> pathlib.Path(b'foo', b'bar')
TypeError: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'bytes'
```

This message is better, and should be used in the former case.

### Linked PRs
* gh-103526

**Comment by terryjreedy** [comment]
I was expecting the same 'confusing message as above for the following, but...
```python
>>> pathlib.Path('foo', 1)
Traceback (most recent call last):
  File "C:\Programs\Python312\Lib\pathlib.py", line 703, in __init__
    super().__init__(*args)
  File "C:\Programs\Python312\Lib\pathlib.py", line 279, in __init__
    path = self._flavour.join(*args)
  File "<frozen ntpath>", line 149, in join
  File "<frozen genericpath>", line 164, in _check_arg_types
TypeError: join() argument must be str, bytes, or os.PathLike object, not 'int'
```
It would be nicer if this also got the improved message about Path args.

**Comment by barneygale** [comment]
That's possible to fix, but it involved either capturing and re-raising the TypeError, or basically re-implementing `os.fspath()`. This code is pretty performance-sensitive so I don't feel great about it.

Another option would be to add a `string_only` argument to `os.fspath()` that rejects bytes and raises an appropriate exception.

**Comment by barneygale** [comment]
> That's possible to fix, but it involves either capturing and re-raising the TypeError, or basically re-implementing `os.fspath()`. This code is pretty performance-sensitive so I don't feel great about it

^ nevermind, I'm wrong. Thanks for the suggestion on the PR!

**Comment by barneygale** [comment]
Fixed in #103526 / https://github.com/python/cpython/commit/8611e7bf5ceace998fefcbf26ab1c5d5bc8a0e2a. Not backported as the bug only affects 3.12.
